contracts/InnerMajoritySet.sol

```diff
@@ -0,0 +1,299 @@
+	}
+
+	// Track the first benign misbehaviour.
+	function firstBenign(address validator) private has_not_benign_misbehaved(validator) {
+		validatorsStatus[validator].firstBenign[msg.sender] = now;
+	}
+
+	// Report that a validator has been repeatedly misbehaving.
+	function repeatedBenign(address validator) private has_repeatedly_benign_misbehaved(validator) {
+		AddressVotes.insert(validatorsStatus[validator].benignMisbehaviour, msg.sender);
+		confirmedRepeatedBenign(validator);
+	}
+
+	// When enough long term benign misbehaviour votes have been seen, remove support.
+	function confirmedRepeatedBenign(address validator) private agreed_on_repeated_benign(validator) {
+		validatorsStatus[validator].firstBenign[msg.sender] = 0;
+		AddressVotes.remove(validatorsStatus[validator].benignMisbehaviour, msg.sender);
+		removeSupport(msg.sender, validator);
+	}
+
+	// Absolve a validator from a benign misbehaviour.
+	function absolveFirstBenign(address validator) public has_benign_misbehaved(validator) {
+		validatorsStatus[validator].firstBenign[msg.sender] = 0;
+		AddressVotes.remove(validatorsStatus[validator].benignMisbehaviour, msg.sender);
+	}
+
+	// PRIVATE UTILITY FUNCTIONS
+
+	// Add a status tracker for unknown validator.
+	function newStatus(address validator) private has_no_votes(validator) {
+		validatorsStatus[validator] = ValidatorStatus({
+			isValidator: false,
+			index: pendingList.length,
+			support: AddressVotes.Data({ count: 0 }),
+			supported: new address[](0),
+			benignMisbehaviour: AddressVotes.Data({ count: 0 })
+		});
+	}
+
+	// ENACTMENT FUNCTIONS (called when support gets out of line with the validator list)
+
+	// Add the validator if supported by majority.
+	// Since the number of validators increases it is possible to some fall below the threshold.
+	function addValidator(address validator) public is_not_validator(validator) has_high_support(validator) {
+		validatorsStatus[validator].index = pendingList.length;
+		pendingList.push(validator);
+		validatorsStatus[validator].isValidator = true;
+		// New validator should support itself.
+		AddressVotes.insert(validatorsStatus[validator].support, validator);
+		validatorsStatus[validator].supported.push(validator);
+		initiateChange();
+	}
+
+	// Remove a validator without enough support.
+	// Can be called to clean low support validators after making the list longer.
+	function removeValidator(address validator) public is_validator(validator) has_low_support(validator) {
+		uint removedIndex = validatorsStatus[validator].index;
+		// Can not remove the last validator.
+		uint lastIndex = pendingList.length-1;
+		address lastValidator = pendingList[lastIndex];
+		// Override the removed validator with the last one.
+		pendingList[removedIndex] = lastValidator;
+		// Update the index of the last validator.
+		validatorsStatus[lastValidator].index = removedIndex;
+		delete pendingList[lastIndex];
+		pendingList.length--;
```

Comment:
this will actually delete the last element in the `pendingList`, so you don't need the line above